### ISSUE TITLE
Support Sass module syntax

### ISSUE THREAD

**Issue by mirisuzanne** [oddbird/accoutrement#43]
See:
- https://github.com/sass/sass/blob/master/accepted/module-system.md
- http://modules.sass-lang.com/
- npm `sass@1.23.0-module.beta.1`

This will be a breaking change, and require a major release.

**Comment by mirisuzanne** [comment]
I've run into some issues with this upgrade, and wanted to document my thoughts for how to handle this.

The parser currently has a some strange recursive dependencies. Both `$functions` & `$ratios` maps are used by the parser, while also needing to be parsed themselves. On one hand, we could probably break that loop by making "adjustments" unavailable in those two maps. On the other hand… I wonder if Sass modules change our approach to the very core of this toolkit?

The core Accoutrement map syntax was invented to help avoid several limitations of working with Sass variables & maps natively:

- Variables couldn't be created or accessed algorithmically.
- Map values couldn't reference other keys in the same map for normal functional adjustments.

Using Sass modules, we can work around some of those issues natively now by accessing the variables in a module as _either_ individual variables, or as a single auto-generated map using `meta.module-variables('module-name')`. 

In some ways, that allows a "best of both worlds" approach, where:

- Initial tokens can be defined together in a single module, using native variables/functions for relationships
- Other modules can access a map of those variables directly or algorithmically, looping over them as desired
- Those maps can be merged together into something like a global Accoutrement map - where some of the tokens are defined by variables in a module, and others can be added algorithmically

The main downside is that this _can lead to an overload of small individual modules_. But the advantage is, we're not maintaining a totally unique value-manipulation syntax inside Sass. We could instead focus on tools to help smooth out that flow. Now that Sass has built-in functions for working with nested maps, I think the creation flow is good. Probably the most useful feature would be allowing you to easily filter maps?

And, of course, we would continue to provide plugins/helpers for colors, type, layouts, etc.

**Comment by mirisuzanne** [comment]
At this point the OddBird team is planning to move towards the Sass modules "best of both worlds" approach, and write up the migration path for that. That might call for some new map helpers (a `filter()` function?), but would allow us to remove the old map syntax & parser, dramatically simplifying the tool. This would be a major change, but we think it's better to help move people into native Sass features where we can – and use Accoutrement to fill in gaps only where needed. Meanwhile, of course, people would be welcome to lock the version and continue to use the old syntax indefinitely on legacy projects. The new major release would only be for versions of Sass that support the module syntax.

We'd be happy for feedback on that from other active users.

/cc @johncrim @joelschou @jcamilleri13 @schliesser @waako if any of you are interested in weighing in

**Comment by jcamilleri13** [comment]
Thanks for the shoutout! At this point, given the advancements in the core Sass libraries, accoutrement would benefit from a slightly leaner approach to the map handling, where the bulk of that work could be offloaded to the core map functions. Now that there's finally support for nested maps on the horizon - hooray - the core accoutrement library can benefit from the speed of the internal functions and lose a pile of complexity in the process.

I personally found one of the most powerful features of accoutrement to be the the ability to reference "internal" map values, i.e. having a value point to another deeply nested value, which allowed for easy customisation of complex systems. Is this something that would still be possible with the new structure  you have in mind? In either case though, off-loading functionality to the Sass core where it can be done is always a preferable solution I think.

**Comment by mirisuzanne** [comment]
@jcamilleri13 The proposal would be to remove (or at least dramatically simplify) how that internal map reference works. The proposal is to replace something like:

```
$colors: (
  // set explicit colors
  'brand-pink': hsl(330, 85%, 62%),
  'brand-light': #ddf,
  'brand-dark': #224,

  // reference existing colors
  'background': '#brand-light',
  'border': '#brand-dark',

  // make adjustments as needed, using color functions
  'link': '#brand-pink' ('shade': 25%, 'desaturate': 15%),
);
```

With an approach based in native Sass modules, which allow you to treat an entire file of variables…

```
// _colors.scss
// ------------
@use 'sass:color';
@use 'accoutrement/tools';

// set explicit colors
$brand-pink: hsl(330, 85%, 62%);
$brand-light: #ddf;
$brand-dark: #224;

// reference existing colors
$background: $brand-light;
$border: $brand-dark;

// make adjustments as needed, using color functions
$link: color.desaturate(tools.shade($brand-pink, 25%), 15%);
```

As _either/botth_ a set of individual variables, or as a map:

```scss
@use 'sass:meta';
@use 'colors';

// access individual colors as variables
a { color: colors.$link; }

// or loop over a map of the colors
@each $name, $color in meta.module-variables('colors') {
  [data-color="#{$name}"] { color: $color }
}
```

You can already merge several maps together, which would allow combining multiple modules one larger `$colors` map, with nested submaps as desired. But the references between keys are all handled using native variables & functions. If we add a good `filter()` option, it could help go the other way as well, breaking large modules into smaller maps when needed. 

For our use-cases, that means we basically plan to convert our old Accoutrement self-referential maps into distinct modules, where we can use normal variables & functions for the references/adjustments, and still get the benefits of a map on the other end, using `meta.module-variables()`.

**Comment by jcamilleri13** [comment]
Looks like a great solution. Large maps tended to get unwieldy, I generally split it into separate files and then import them all into one big map to help organise it better. Sounds great 🥳

**Comment by dvdherron** [comment]
My thoughts after going through and taking an inventory of the things I use or find useful. 

Core:
I think a lot of this will end up being split off into their own modules?
- Keep: CSS Variables, Ratios & Modular Scales

I can't say I've used so much of the Core/Utilities, but I think these were mostly used in the parser? I think they could mostly be split off into separate modules if we keep them. 

Init (Normalization)
- As we discussed, I think it's great this comes with a way to reset styles. We tend to use CSS-Remedy more, so maybe this could all be replaced with that?

Plugin/Animate
- Keep. I've only touched the surface of using some of these but they help with cutting down on how much code one has to write. 

Plugin/Color
- Keep

Plugin/Layout
- Keep

Scale
- Keep

Type
- Keep/change?. I will have to revisit some of our conversation on setting up web fonts in one of our projects (TD). There might be some things we talked about possibly changing here.

**Comment by johncrim** [comment]
Thank you very much for looping me in. I particularly appreciate that you're considering community feedback even though this project was originally written for your consulting practice.

I've had to give this proposal some thought before responding: 

I understand and agree that the parser has weird recursive dependencies - I've commented on that before, and I think parsing would benefit from a rewrite (I can envision how it could be much simpler, particularly given the nested keys in the new `map.get` -- and I'm happy to help, though I haven't done much yet). I also am a fan of using more native sass APIs and generally re-aligning accoutrement with the new sass APIs, and making accoutrement leaner in the process.

The part I disagree with is switching from primarily using maps to primarily using variables in files (if I understood you right @mirisuzanne). I think the current accoutrement developer API is pretty nice, and I think switching (if that's the intent) to variables in files would be a step backwards. Just because we can use the new sass APIs to view file variables as a map doesn't mean that's the best choice for users of accoutrement.

To explain my feedback more:

We use accoutrement to organize design token definitions and their dependencies, and use functionality that uses those tokens. We use nested maps extensively, and we override values for themes using a map deep merge function that I wrote (it looks like the new `map.deep-merge` in `sass:map` does the same thing). I've written additional "modules" similar to the built-in accoutrement modules using the same pattern - each module has a global map specific to the domain and functions/mixins that use the map. In addition to extending the built-in colors and animations modules, I've written accoutrement-like modules for these domains: 
	• Gradients
	• Opacity
	• Elevation
	
We've been using accoutrement with sass modules for almost 2 years, and most of the accoutrement API works well as-is with sass modules.

Things we value that may be difficult using plain variables in files instead of maps:
	• Convenient + concise named references that are grouped by domain and have an implied type  - eg `a.color('button->primary')`, `a.size('button->m->height')`
	• Multiple maps per file (eg theme definition includes size, color, and animation overrides)
	• Merging maps (eg: defaults + dark-theme) that are defined in different files.
	• Lazy evaluation - it can be nice to defer evaluating values until they are used. Would also be nice to have a switch to enable eager evaluation for debugging.
	• Functions can generate new definition maps from existing definition maps

Pros/cons of files vs maps:
	• Files variables, when viewed as a map, are flat (1 level) - for hierarchical configuration, one would need to use maps within files
	• Can use built-in functions in both files and maps
	• "Private" variables can be used in both
	• Multiple sets of maps can be defined within a file; File level configuration would require one file per domain
	• Merging + overriding maps is easier to reason about than merging file variables (I think)

Net-net, I'm not a fan of switching to files as the default container for configuration; I am however a fan of supporting it as an option. Using files as the default container, instead of maps, may improve interoperability with projects that don't use accoutrement, but if used as the default pattern is liable to make configuration harder to keep organized.

**Comment by mirisuzanne** [comment]
Thanks @johncrim - that's a useful rundown of how you use this. I did some work re-writing the parser last week (in PR #76, which is still a bit of a mess). It resolves some of the recursive dependencies by:

- not supporting the special map-parsing syntax inside the `$functions` map
- not supporting ratios as "first-class" akin to functions in the syntax
- I also removed the "silent" option for `$handle-missing-keys`, since we support escaping/ending keys with `\\` (you can still use `null`, `error`, or `warn`
- (I think that's about it for major changes, but I may have missed something here. I haven't finished registering all the built-in functions…)

If something like that looks good, I think we could consider shipping with the parser? Maybe move it into a `tokens` module, and consider it a bit less "core" to the whole project, but still supported?

**Comment by stacyk** [comment]
Here are my initial thoughts.

## Keep
**Layout**
* Media queries (above, below, between)
* Z-index
* Position (sometimes, only when I remember)
* Pseudos Before/After mixins are great

**Scale**
* Unit conversion

**Type** 
* is-hidden

**Animate**
- Everything in here is useful if I remember to use it.
- I had no idea that we had a built in set of easing functions. I'd love to add a demo to show what each looks like in practice.

**Color**
- I love everything in colors 
- I have never used stripe-colors mixin, is that supposed to be in the "themes" group? Is that what is used for the Blend display on Herman?
- Sometimes I prefer to edit the HSL values 

## MAYBE or CUT
**Layout**
- Do we use [fluid-ratio](https://www.oddbird.net/accoutrement/docs/layout-ratios#mixin--fluid-ratio)?
- I'm not sure about how much we overlap with Remedy in our Init plugins (I know quite a bit) but I think we have more accessibility helpers in Init than remedy has, right? 

## ADD? 
- For Top Dog I wanted to target width and height in a media query. I don't think our current options allow for that combination, do they?
- I wonder if the [is-hidden mixin](https://www.oddbird.net/accoutrement/docs/type-a11y#mixin--is-hidden) should move to the [Init Accessibility group](https://www.oddbird.net/accoutrement/docs/init-a11y)

**Comment by patriciarealini** [comment]
### Suggestions for Modules to Remove

8/23 Meeting Notes

Init 
- Remove, in favor of using Remedy 

Layout 
- global-box-sizing
- clear-fix
- fluid-ratio 

Keep everything else

**Comment by mirisuzanne** [comment]
We just released [v4.0.0-beta.2](https://github.com/oddbird/accoutrement/releases/tag/v4.0.0-beta.2), and published that on npm with the `pre` tag. I'd love to have some other folks testing this in real projects before we do the final release. Release notes / changelog should cover major changes. You can also pull and load the docs from source, I'll try and get that deploy preview working as well.

- The PR for it is #94

**Comment by johncrim** [comment]
We've been using [v4.0.0-beta3](https://github.com/oddbird/accoutrement/tree/v4.0.0-beta.3) for about a month, and it's been working great for us. I'm happy with the updated design, and it's noticeably faster. It's also great that a few bugs that I'd had to live with for a while are gone; in particular that `null` is now returned when a token is missing.

I diffed all our CSS output before and after upgrading to v4.0.0 (which was a great way to find any bugs in the migration; I also found a couple bugs in our code). The only change I found is that some of our mixed colors are lighter - presumably that's due to changes in the `sass:color` functions (eg color.scale() vs color.adjust()). I have not yet determined the cause of this change, but I'm also not sure that it's wrong - it might have been wrong before :) . As an engineer I'm somewhat embarrassed to not have found the true cause, but it wasn't worth the time.

The other thing I'll note is that it's definitely a breaking change - probably a majority of the lines of code that use accoutrement had to be touched. So don't tackle it until you have the time.

I'm a fan! This is one vote for "ship it!". @mirisuzanne - thank you very much for your time on this fantastic project and the shiny new version. The only thing I don't understand is why everyone isn't using it. :)
